src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/RoutePatternUsageDetector.cs

```diff
@@ -0,0 +1,200 @@
+                var classSymbol = semanticModel.GetDeclaredSymbol(classDeclarationSyntax, cancellationToken);
+
+                return new(null, false, MvcDetector.IsController(classSymbol, semanticModel));
+            }
+        }
+
+        return default;
+    }
+
+    private static SyntaxNode? FindAttributeParent(SyntaxNode container)
+    {
+        var argument = container.Parent;
+        if (argument.Parent is not AttributeArgumentListSyntax argumentList)
+        {
+            return null;
+        }
+
+        if (argumentList.Parent is not AttributeSyntax attribute)
+        {
+            return null;
+        }
+
+        if (attribute.Parent is not AttributeListSyntax attributeList)
+        {
+            return null;
+        }
+
+        return attributeList.Parent;
+    }
+
+    private static IMethodSymbol? FindMvcMethod(SemanticModel semanticModel, MethodDeclarationSyntax methodDeclaration, CancellationToken cancellationToken)
+    {
+        var methodSymbol = semanticModel.GetDeclaredSymbol(methodDeclaration, cancellationToken);
+
+        if (methodSymbol.ContainingType is not ITypeSymbol typeSymbol)
+        {
+            return null;
+        }
+
+        if (!MvcDetector.IsController(typeSymbol, semanticModel))
+        {
+            return null;
+        }
+
+        if (!MvcDetector.IsAction(methodSymbol, semanticModel))
+        {
+            return null;
+        }
+
+        return methodSymbol;
+    }
+
+    private static IMethodSymbol? FindMapMethod(SemanticModel semanticModel, SyntaxNode container, CancellationToken cancellationToken)
+    {
+        var argument = container.Parent;
+        if (argument.Parent is not BaseArgumentListSyntax argumentList ||
+            argumentList.Parent is null)
+        {
+            return null;
+        }
+
+        // Multiple overloads could be resolved, e.g. MapGet(string, RequestDelegate) and MapGet(string, Delegate)
+        // Check each overload result to see whether it matches and return the first valid result.
+        var symbols = GetBestOrAllSymbols(semanticModel.GetSymbolInfo(argumentList.Parent, cancellationToken));
+
+        foreach (var symbol in symbols)
+        {
+            if (symbol is IMethodSymbol methodSymbol)
+            {
+                var matchingMapSymbol = FindValidMapMethod(semanticModel, argumentList, methodSymbol, cancellationToken);
+                if (matchingMapSymbol != null)
+                {
+                    return matchingMapSymbol;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private static IMethodSymbol? FindValidMapMethod(SemanticModel semanticModel, BaseArgumentListSyntax argumentList, IMethodSymbol method, CancellationToken cancellationToken)
+    {
+        if (!method.Name.StartsWith("Map", StringComparison.Ordinal))
+        {
+            return null;
+        }
+
+        var delegateSymbol = semanticModel.Compilation.GetTypeByMetadataName("System.Delegate");
```

Comment:
there shoudl be a GetSpecialType(SpecialType.System_Delegate) (or something like that).